lib/core/test/unit/Cardano/Wallet/Primitive/Migration/PlanningSpec.hs

```diff
@@ -0,0 +1,367 @@
+spec :: Spec
+spec = describe "Cardano.Wallet.Primitive.Migration.PlanningSpec" $
+
+    modifyMaxSuccess (const 1000) $ do
+
+    parallel $ describe "Creating migration plans" $ do
+
+        describe "Small migrations" $
+            it "prop_createPlan_small" $
+                property prop_createPlan_small
+
+        describe "Large migrations" $
+            it "prop_createPlan_large" $
+                property prop_createPlan_large
+
+        describe "Giant migrations" $
+            it "prop_createPlan_giant" $
+                property prop_createPlan_giant
+
+    parallel $ describe "Categorizing UTxO entries" $ do
+
+        it "prop_categorizeUTxOEntry" $
+            property prop_categorizeUTxOEntry
+
+--------------------------------------------------------------------------------
+-- Creating migration plans
+--------------------------------------------------------------------------------
+
+data ArgsForCreatePlan = ArgsForCreatePlan
+    { mockConstraints :: MockTxConstraints
+    , mockInputs :: [(MockInputId, TokenBundle)]
+    , mockRewardWithdrawal :: Coin
+    }
+    deriving (Eq, Show)
+
+newtype Small a = Small { unSmall :: a }
+    deriving (Eq, Show)
+
+newtype Large a = Large { unLarge :: a }
+    deriving (Eq, Show)
+
+newtype Giant a = Giant { unGiant :: a }
+    deriving (Eq, Show)
+
+instance Arbitrary (Small ArgsForCreatePlan) where
+    arbitrary = Small <$> genArgsForCreatePlan
+        (0, 100) genMockInput
+
+instance Arbitrary (Large ArgsForCreatePlan) where
+    arbitrary = Large <$> genArgsForCreatePlan
+        (1_000, 1_000) genMockInput
+
+instance Arbitrary (Giant ArgsForCreatePlan) where
+    arbitrary = Giant <$> genArgsForCreatePlan
+        (100_000, 100_000) genMockInputAdaOnly
+
+prop_createPlan_small :: Blind (Small ArgsForCreatePlan) -> Property
+prop_createPlan_small (Blind (Small args)) =
+    withMaxSuccess 1000 $
+    prop_createPlan args
+
+prop_createPlan_large :: Blind (Large ArgsForCreatePlan) -> Property
+prop_createPlan_large (Blind (Large args)) =
+    withMaxSuccess 100 $
+    prop_createPlan args
+
+prop_createPlan_giant :: Blind (Giant ArgsForCreatePlan) -> Property
+prop_createPlan_giant (Blind (Giant args)) =
+    withMaxSuccess 1 $
+    prop_createPlan args
+
+genArgsForCreatePlan
+    :: (Int, Int)
+    -- ^ Input count range
+    -> (MockTxConstraints -> Gen (MockInputId, TokenBundle))
+    -- ^ Genenator for inputs
+    -> Gen ArgsForCreatePlan
+genArgsForCreatePlan (inputCountMin, inputCountMax) genInput = do
+    mockConstraints <- genMockTxConstraints
+    mockInputCount <- choose (inputCountMin, inputCountMax)
+    mockInputs <- replicateM mockInputCount (genInput mockConstraints)
+    mockRewardWithdrawal <- oneof
+        [ pure (Coin 0)
+        , genCoinRange (Coin 1) (Coin 1_000_000)
+        ]
+    pure ArgsForCreatePlan
+        { mockConstraints
+        , mockInputs
+        , mockRewardWithdrawal
+        }
+
+prop_createPlan :: ArgsForCreatePlan -> Property
```

Comment:
I wonder: isn't this `ArgsX` pattern similar to `forAll` 🤔 

Haven't done this myself, but I think you could have 
```haskell
prop_createPlan :: MockTxConstraints -> MockWithdrawal -> Property
prop_createPlan constraints (MockWithdrawal wdrl) = forAll (genMockInputs constraints) $ \inputs -> 
```
such that the generators are only _as coupled_ as necessary but no more, and you can more easily have shrinkers for the constraints and the withdrawal.

Reply:
Hi @Anviking 

As requested, I've used `forAll` and `forAllShrink` to replace all the `ArgsX` types.

Thanks for the suggestion, I think it has really simplified the test code. (Much less boilerplate!)

Reply:
Nice! 🎉